tests: bluetooth: classic: sim: add GAP discovery test

lylezhu2012 opened the pull request:
Add Bluetooth Classic GAP (Generic Access Profile) discovery test for the simulation testing infrastructure.

The test validates inquiry-based device discovery functionality, including both general and limited discovery procedures. It uses two simulated Bumble controllers connected via TCP to test central and peripheral roles performing discovery, connection establishment, and disconnection.

It is the first test suite based on the Classic simulation environment.

lylezhu2012 commented on tests/bluetooth/classic/gap_discovery/src/test_central.c:
```diff
@@ -0,0 +1,231 @@
+	err = bt_br_discovery_stop();
+	if (err != 0) {
+		LOG_ERR("Failed to stop GAP discovery procedure (err %d)", err);
+	}
```
The error code of GAP discovery stop is ignored. Just add a log to print it.

jhedberg commented on .github/workflows/bt_classic_sim_tests.yml:
```diff
@@ -0,0 +1,215 @@
+          wget https://github.com/zephyrproject-rtos/sdk-ng/releases/download/v0.16.5/zephyr-sdk-0.16.5_linux-x86_64.tar.xz
+          wget -O - https://github.com/zephyrproject-rtos/sdk-ng/releases/download/v0.16.5/sha256.sum | shasum --check --ignore-missing
+          tar xvf zephyr-sdk-0.16.5_linux-x86_64.tar.xz
+          cd zephyr-sdk-0.16.5
```
Why such an old version of the SDK? The latest one is `0.17.4`. You might also want to factor out the version to a single updatable variable, and maybe also consider using the `west sdk install` command.

aescolar reviewed (dismissed):
I have only done a quick very partial review. But so far:
Questions:
* Is this setup deterministic? will results depend on the machine load?
* Can this setup run tests in parallel? (or are we limited to run tests serially independently of the number of cores in the machine) (EDIT: PR was changed to parallelize tests)

Fundamental issue:
* If we are going to have another tests setup, we should have some documentation for it in Zephyr's rst.

lylezhu2012 commented on tests/bluetooth/classic/sim/gap_discovery/tests_scripts/gap_discovery_run.sh:
```diff
@@ -0,0 +1,595 @@
+        return 1
+    fi
+
+    log_info "zephyr.exe found successfully at $zephyr_exe"
```
Removed.

lylezhu2012 commented on .github/workflows/bt_classic_sim_tests.yml:
```diff
@@ -0,0 +1,151 @@
+
+      - name: Create Test Results
+        run: |
+          junit2html tests/bluetooth/classic/sim/test_logs/junit.xml tests/bluetooth/classic/sim/test_logs/junit.html
```
Put all log files to the same folder `bt_classic_sim`.

lylezhu2012 commented on .github/workflows/bt_classic_sim_tests.yml:
```diff
@@ -0,0 +1,151 @@
+      - name: Cleanup
+        if: always()
+        run: |
+          sudo pkill -9 zephyr.exe || true
```
Removed.

lylezhu2012 commented on tests/bluetooth/classic/sim/ci.bt.classic.sh:
```diff
@@ -0,0 +1,507 @@
+    cd "$original_dir" 2>/dev/null || true
+
+    # Record end time
+    local end_time=$(date +%s)
```
Change the unit to `ms`.

lylezhu2012 commented on tests/bluetooth/classic/sim/ci.bt.classic.sh:
```diff
@@ -0,0 +1,507 @@
+    local total_tests=$((${#PASSED_TESTS[@]} + ${#FAILED_TESTS[@]}))
+    local total_time=0
+
+    # Calculate total time
```
Added. Calculated the total time separately.

lylezhu2012 commented on tests/bluetooth/classic/sim/ci.bt.classic.sh:
```diff
@@ -0,0 +1,507 @@
+    log_info "CI BT Classic cleaning up..."
+
+    # Kill any remaining processes (try with and without sudo)
+    pkill -9 btvirt 2>/dev/null || sudo pkill -9 btvirt 2>/dev/null || true
```
Removed.

lylezhu2012 commented on tests/bluetooth/classic/sim/ci.bt.classic.sh:
```diff
@@ -0,0 +1,507 @@
+    # Kill any remaining processes (try with and without sudo)
+    pkill -9 btvirt 2>/dev/null || sudo pkill -9 btvirt 2>/dev/null || true
+    pkill -9 btmon 2>/dev/null || sudo pkill -9 btmon 2>/dev/null || true
+    pkill -9 zephyr.exe 2>/dev/null || sudo pkill -9 zephyr.exe 2>/dev/null || true
```
All of them are removed.

lylezhu2012 commented on tests/bluetooth/classic/sim/gap_discovery/tests_scripts/gap_discovery_run.sh:
```diff
@@ -0,0 +1,595 @@
+    display_controller_info
+
+    # 3. Build peer device
+    if ! build_peer_device; then
```
The script do not support this case.

The BD address of simulation controller is randomly generated.

And the precondition of building includes the peer BD address in current test case.

Therefore, when creating a simulation controller, the build and run processes should be executed sequentially. This ensures that the addresses are not modified.

lylezhu2012 commented on doc/develop/test/bluetooth_classic_sim.rst:
```diff
@@ -0,0 +1,977 @@
+The ``tests.native_sim.txt`` file serves as a registry of all Bluetooth Classic simulation
+tests that should be included in the build and test execution process. This allows:
+
+- **Centralized test management**: All tests are listed in one location
```
It is updated.

lylezhu2012 commented:
> We do it with our scripting. In the case of the twister job, it is called again (up to 3 times in total) told to rerun only last run failed tests. An option would be for this PR parallelizing script to retry up to N times if a test fails on its own.

Thanks. Added an argument MAX_RUN_RETRIES for retry.

chengkai15 commented on tests/bluetooth/classic/sim/common/bt_sim_controller.py:
```diff
@@ -0,0 +1,349 @@
+        self.count = count
+        self.interval = interval
+        if on_inquiry:
+            self.on_complete = on_inquiry
```
```suggestion
            self.on_inquiry = on_inquiry
```

chengkai15 commented:
@lylezhu2012 any update?

nashif commented:
given the recent changes in twister to support running babblesim tests (using script harness: https://github.com/zephyrproject-rtos/zephyr/pull/108119), was there any attempt to make this type of tests run and be managed by twister? 
It would be great to have testing all done in one worflow and managed the same way, having multiple workflows with different envionment setups that are managed by different folks is not ideal.
But in general I am not opposed for giving this a try short term with the goal of consolidating our workflows and how we run and manage tests mid to long term.

lylezhu2012 commented:
> given the recent changes in twister to support running babblesim tests (using script harness: #108119), was there any attempt to make this type of tests run and be managed by twister? It would be great to have testing all done in one worflow and managed the same way, having multiple workflows with different envionment setups that are managed by different folks is not ideal. But in general I am not opposed for giving this a try short term with the goal of consolidating our workflows and how we run and manage tests mid to long term.

Thanks for sharing. Let me try it.

lylezhu2012 committed:
docs: bluetooth: classic: add simulation testing framework doc

Add comprehensive documentation for the Bluetooth Classic simulation
testing framework that enables testing BR/EDR functionality without
physical hardware.

Use GAP discovery as a reference, demonstrate how to add a new test
suites for Bluetooth Classic Simulation.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
tests: bluetooth: classic: sim: add common utilities

Add common utility scripts and Python modules for Bluetooth Classic
simulation testing infrastructure.

The added utilities provide the following auxiliary functions for
designing scripts to simulate tests for test suites: obtaining the
executable file compiled by Twister, installing the plugin
`Bumble 0.0.220`, creating and managing virtual controllers, creating
test threads, and waiting for test threads to complete.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
tests: bluetooth: classic: sim: add GAP discovery test

Add Bluetooth Classic GAP (Generic Access Profile) discovery test
for the simulation testing infrastructure.

The test validates inquiry-based device discovery functionality,
including both general and limited discovery procedures. It uses
two simulated Bumble controllers connected via TCP to test central
and peripheral roles performing discovery, connection establishment,
and disconnection.

It is the first test suites based on the Classic simulation env.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
tests: bt: classic: sim: gap_discovery: config limited disc duration

Configure BT_LIMITED_DISCOVERABLE_DURATION to 3 seconds for the GAP
discovery simulation test to reduce test execution time.

The configuration limits the duration range to 1-60 seconds and is
specific to this test suite only.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 commented:
Hi @nashif, @aescolar, @jhedberg, @Thalley, @chengkai15 

I have removed the changes related to the GitHub workflow and updated the PR using the `twister + harness + script` testing framework.

Please help review the changes.

chengkai15 commented on tests/bluetooth/classic/sim/gap_discovery/src/test_central.c:
```diff
@@ -0,0 +1,244 @@
+	peer_device_connect();
+}
+
+#define BT_COD_MAJOR_SVC_CLASS_LIMITED_DISCOVER BIT(13)
```
BT_COD_MAJOR_SVC_CLASS_LIMITED_DISCOVER is defined in
include/zephyr/bluetooth/assigned_numbers.h

chengkai15 commented on tests/bluetooth/classic/sim/gap_discovery/src/test_central.c:
```diff
@@ -0,0 +1,244 @@
+};
+
+/* Multiplier for each discovery result to estimate timeout (seconds per device) */
+#define GAP_DISCOVERY_TIMEOUT_PER_DEVICE 1.25
```
it seems that the length was in units of 1.28 s

lylezhu2012 replied:
Corrected.

chengkai15 commented on tests/bluetooth/classic/sim/gap_discovery/src/test_central.c:
```diff
@@ -0,0 +1,244 @@
+#define GAP_DISCOVERY_TIMEOUT_PER_DEVICE 1.25
+/* Base timeout added to the total (seconds) */
+#define GAP_DISCOVERY_TIMEOUT_BASE 5.0
+#define GAP_DISCOVERY_TIMEOUT(count) \
```
For the Z_TIMEOUT_MS / K_MSEC / SYS_CLOCK series, suggest using integer milliseconds for time calculations is the standard convention in Zephyr.

lylezhu2012 replied:
Corrected.

chengkai15 commented on tests/bluetooth/classic/sim/common/common_utils.sh:
```diff
@@ -0,0 +1,593 @@
+# Args: script_list (space-separated or newline-separated)
+#       max_jobs (optional, default: number of CPU cores)
+# Returns: 0 if all scripts succeed, 1 if any fail
+run_parallel() {
```
run_parallel/collect_parallel_results — hand-assembles JUnit <testsuites> XML, duplicating what the Twister script harness already provides.

lylezhu2012 replied:
Removed.

chengkai15 commented on tests/bluetooth/classic/sim/common/thread_utils.sh:
```diff
@@ -0,0 +1,140 @@
+# Args: command_line
+# Returns: PID of the background process
+execute_async() {
+    local command_line="$*"
```
uses eval "$*", which would re-parses arguments and loses quoting

lylezhu2012 replied:
Fixed.

chengkai15 commented:
I haven't gone through everything yet — it's a fairly large change. 
 
 But based on the discussion so far, it seems we're not fully aligned on the overall approach yet. 
 
 Could we settle the framework direction (extend bsim vs. adopt Bumble) before going further?

lylezhu2012 commented on tests/bluetooth/classic/sim/gap_discovery/Kconfig:
```diff
@@ -0,0 +1,15 @@
+
+# Only for use with the Bluetooth Classic GAP Discovery simulation test
+# Used to shorten test time by limiting discovery duration
+config BT_LIMITED_DISCOVERABLE_DURATION
+	default 3
+	range 1 60
```
The purpose is to modify the range. The type is not needed.

lylezhu2012 commented on tests/bluetooth/classic/sim/common/common_utils.sh:
```diff
@@ -0,0 +1,593 @@
+                exit 1
+            fi
+
+            if [[ $? -eq 0 && -n "${resolved}" ]]; then
```
Cleared the code block

lylezhu2012 commented on tests/bluetooth/classic/sim/common/common_utils.sh:
```diff
@@ -0,0 +1,593 @@
+
+            if [[ $? -eq 0 && -n "${resolved}" ]]; then
+                valid_dirs+=("${resolved}")
+            fi
```
Cleared the code block

lylezhu2012 commented on tests/bluetooth/classic/sim/common/thread_utils.sh:
```diff
@@ -0,0 +1,140 @@
+    fi
+
+    return 0
+}
+
+# Function to execute command asynchronously
+# Args: command_line
+# Returns: PID of the background process
+execute_async() {
+    local command_line="$*"
+    local array_name=$(get_pids_array_name)
+
+    if [[ -z "$command_line" ]]; then
+        echo "Error: No command provided to execute_async" >&2
+        return 1
+    fi
+
+    eval "$command_line" &
+    local pid=$!
+
+    if ! kill -0 "$pid" 2>/dev/null; then
+        echo "Error: Failed to start async process" >&2
+        return 1
+    fi
+
+    # Save PID to process-specific array
+    eval "${array_name}+=('$pid')"
+
+    echo "$pid"
+    return 0
+}
```
~Ignored and unchanged.~

Fixed.

lylezhu2012 commented on tests/bluetooth/classic/sim/gap_discovery/src/test_central.c:
```diff
@@ -0,0 +1,244 @@
+				    ARRAY_SIZE(br_discover_result));
+	zassert_equal(err, 0, "Bluetooth inquiry failed (err %d)", err);
+
+	/* Wait for all discovery results to be processed */
+	timeout = GAP_DISCOVERY_TIMEOUT(ARRAY_SIZE(br_discover_result));
+	LOG_DBG("Will wait for GAP discovery done (timeout %us)", (uint32_t)timeout);
+	err = k_sem_take(&br_discover_sem, K_SECONDS(timeout));
+	zassert_equal(err, 0, "Failed to wait for discovery done (err %d)", err);
```
Fixed.

lylezhu2012 commented on tests/bluetooth/classic/sim/gap_discovery/README.rst:
```diff
@@ -0,0 +1,95 @@
+* starts two Bumble controllers (one per port, with fixed BD_ADDR values),
+* copies the generated ``zephyr.exe`` into ``${BT_CLASSIC_BIN}`` using
+  ``bt_classic_${BOARD_SANITIZED}_bluetooth.classic.gap.discovery.exe`` naming,
+* launches the test binary twice:
```
Fixed

lylezhu2012 commented on doc/develop/test/bluetooth_classic_sim.rst:
```diff
@@ -0,0 +1,880 @@
+       "$HCI_PORT_1@${BD_ADDR[1]}")
+
+   test_case_name="bluetooth.classic.gap.discovery"
+   extra_twister_out "${ZEPHYR_BASE}/twister-out" "${test_case_name}"
```
Fixed.

lylezhu2012 commented on doc/develop/test/bluetooth_classic_sim.rst:
```diff
@@ -0,0 +1,880 @@
+   test_exe="${BT_CLASSIC_BIN}/bt_classic_${BOARD_SANITIZED}_${test_case_name}.exe"
+
+   test_case_name="bluetooth.classic.gap.discovery"
+   extra_twister_out "${ZEPHYR_BASE}/twister-out" "${test_case_name}"
```
Fixed

lylezhu2012 commented on tests/bluetooth/classic/sim/common/bt_sim_controller.py:
```diff
@@ -0,0 +1,928 @@
+        '''
+        logger.debug(
+            f'Inqury LAP {command.lap} length {command.inquiry_length} '
+            f'num_responses {command.num_responses}'
```
Fixed.

lylezhu2012 commented on tests/bluetooth/classic/sim/common/common_utils.sh:
```diff
@@ -0,0 +1,593 @@
+    local base_dir="$2"
+    local test_dirs
+    local valid_dirs=()
+
```
Fixed

lylezhu2012 commented on tests/bluetooth/classic/sim/gap_discovery/tests_scripts/limited_discovery.sh:
```diff
@@ -0,0 +1,91 @@
+extra_twister_out "${ZEPHYR_BASE}/twister-out" "${test_case_name}"
+
+peripheral_exe="${BT_CLASSIC_BIN}/bt_classic_${BOARD_SANITIZED}_${test_case_name}.exe"
+ARGS=("--peer_bd_address=${BD_ADDR[1]}" \
+      " -test=gap_peripheral::test_02_gap_peripheral_limited_discovery")
+
```
Fixed

lylezhu2012 commented on tests/bluetooth/classic/sim/gap_discovery/tests_scripts/limited_discovery.sh:
```diff
@@ -0,0 +1,91 @@
+execute_async "$peripheral_exe" --bt-dev="127.0.0.1:$HCI_PORT_0" "${ARGS[@]}" || exit 1
+
+central_exe="${BT_CLASSIC_BIN}/bt_classic_${BOARD_SANITIZED}_${test_case_name}.exe"
+ARGS=("--peer_bd_address=${BD_ADDR[0]}" \
+      " -test=gap_central::test_02_gap_central_limited_discovery")
+
```
Fixed.

lylezhu2012 commented on doc/develop/test/bluetooth_classic_sim.rst:
```diff
@@ -0,0 +1,880 @@
+   # Pass peripheral's address as peer_bd_address parameter
+   central_exe="${BT_CLASSIC_BIN}/bt_classic_${BOARD_SANITIZED}_${test_case_name}.exe"
+   ARGS="--peer_bd_address=${BD_ADDR[0]}  -test=gap_central::test_01_gap_central_general_discovery"
+   execute_async "$central_exe" --bt-dev="127.0.0.1:$HCI_PORT_1" "$ARGS" || exit 1
```
Fixed.

lylezhu2012 commented:
> the overall approach yet.

It is unclear what you mean by "overall approach."

> Could we settle the framework direction (extend bsim vs. adopt Bumble) before going further?

Because Bluetooth Classic is unsupported by BubbleSim, the BSIM cannot be used. So, for Bluetooth Classic, the test framework is ZTEST (twister + harness + script) in my plan.

So, the test framework is clear. It is ZTEST (twister + harness + script). Also it is applicable `BSIM`.

lylezhu2012 commented on tests/bluetooth/classic/sim/common/port_utils.sh:
```diff
@@ -0,0 +1,430 @@
+﻿#!/usr/bin/env bash
```
Fixed,

lylezhu2012 commented on tests/bluetooth/classic/sim/gap_discovery/Kconfig:
```diff
@@ -0,0 +1,15 @@
+
+# Only for use with the Bluetooth Classic GAP Discovery simulation test
+# Used to shorten test time by limiting discovery duration
+config BT_LIMITED_DISCOVERABLE_DURATION
+	default 3
+	range 1 60
```
Not accepted.

lylezhu2012 commented on tests/bluetooth/classic/sim/gap_discovery/CMakeLists.txt:
```diff
@@ -0,0 +1,10 @@
+# SPDX-License-Identifier: Apache-2.0
+
+cmake_minimum_required(VERSION 3.20.0)
+set(NO_QEMU_SERIAL_BT_SERVER 1)
+
+find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE})
+project(bluetooth)
+
+FILE(GLOB app_sources src/*.c)
+target_sources(app PRIVATE ${app_sources})
```
Ignored.

lylezhu2012 commented on tests/bluetooth/classic/sim/common/thread_utils.sh:
```diff
@@ -0,0 +1,142 @@
+    for pid in "${pids_copy[@]}"; do
+        wait $pid || let "exit_code=$?"
+    done
+
```
Optimized.

lylezhu2012 commented on tests/bluetooth/classic/sim/common/common_utils.sh:
```diff
@@ -0,0 +1,89 @@
+if [[ -n "${COMPILE_UTILS_LOADED:-}" ]]; then
+    return 0
+fi
+COMPILE_UTILS_LOADED=1
```
Updated.

lylezhu2012 commented on tests/bluetooth/classic/sim/common/bumble_utils.sh:
```diff
@@ -0,0 +1,166 @@
+        return 0
+    fi
+
+    if [[ -z "$installed_version" ]]; then
+        echo "Bumble not installed; installing Bumble==$required_version"
+    else
+        echo "Bumble version mismatch (installed: $installed_version, " \
+             "required: $required_version); reinstalling"
+        "$py" -m pip uninstall -y bumble >/dev/null 2>&1 || true
+    fi
+
+    # Ensure pip is available
+    if ! "$py" -m pip --version >/dev/null 2>&1; then
+        echo "Error: pip not available for '$py'; cannot install Bumble" >&2
+        return 1
+    fi
+
+    # Install the pinned version
+    if ! "$py" -m pip install --upgrade "Bumble==$required_version"; then
+        echo "Error: Failed to install Bumble==$required_version" >&2
+        return 1
+    fi
```
If the test is executed by a human, this comment is fine; however, for CI, this comment is unacceptable.

lylezhu2012 commented on tests/bluetooth/classic/sim/common/bt_sim_link.py:
```diff
@@ -0,0 +1,255 @@
+                        Class_Of_Device=[c.class_of_device],
+                        Clock_Offset=[0x00],
+                        RSSI=[-60],
+                    )
```
Updated

lylezhu2012 commented on tests/bluetooth/classic/sim/gap_discovery/src/test_central.c:
```diff
@@ -0,0 +1,246 @@
+	double timeout;
+	uint64_t timeout_ms;
+
+	atomic_clear_bit(test_flags, TEST_FLAG_DEVICE_FOUND);
+
+	LOG_DBG("Starting Bluetooth inquiry");
+
+	br_discover_param.length = BR_DISCOVER_RESULT_COUNT;
+	br_discover_param.limited = limited;
+
```
Fixed

lylezhu2012 commented on tests/bluetooth/classic/sim/gap_discovery/src/test_central.c:
```diff
@@ -0,0 +1,246 @@
+	/* Wait for all discovery results to be processed */
+	timeout = GAP_DISCOVERY_TIMEOUT(ARRAY_SIZE(br_discover_result));
+	timeout_ms = (uint64_t)(timeout * 1000.0);
+	LOG_DBG("Will wait for GAP discovery done (timeout %llu ms)", timeout_ms);
```
Optimized.

lylezhu2012 commented on tests/bluetooth/classic/sim/gap_discovery/src/test_main.c:
```diff
@@ -0,0 +1,55 @@
+#include <zephyr/bluetooth/bluetooth.h>
+#include <zephyr/bluetooth/hci.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/ztest.h>
+
+#include "nsi_host_trampolines.h"
+#include "nsi_errno.h"
+#include "soc.h"
+#include "cmdline.h" /* native_sim command line options header */
+
+#define LOG_MODULE_NAME test_gap_discovery_main
+LOG_MODULE_REGISTER(LOG_MODULE_NAME, CONFIG_TEST_GAP_DISCOVERY_LOG_LEVEL);
+
+extern bt_addr_t peer_addr;
+
+static void cmd_peer_bd_address_found(char *argv, int offset)
+{
+	char *addr_str = &argv[offset];
+	int err;
+
+	err = bt_addr_from_str(addr_str, &peer_addr);
+	if (err != 0) {
+		LOG_ERR("Failed to parse peer Bluetooth address: %s (err %d)", addr_str, err);
+	}
+}
```
Added.

jhedberg commented:
@lylezhu2012 I added you back as assignee, since this doesn't touch runner details or anything else outside of the Bluetooth Classic domain.

aescolar commented on tests/bluetooth/classic/sim/gap_discovery/src/test_main.c:
```diff
@@ -0,0 +1,57 @@
+			.option = "peer_bd_address",
+			.name = "XX:XX:XX:XX:XX:XX",
+			.type = 's',
+			.dest = NULL,
```
Nit:
```suggestion
			.is_mandatory = true,
			.option = "peer_bd_address",
			.name = "XX:XX:XX:XX:XX:XX",
			.type = 's',
```
No need to set fields that are 0

aescolar commented on tests/bluetooth/classic/sim/gap_discovery/testcase.yaml:
```diff
@@ -0,0 +1,14 @@
+
+tests:
+  bluetooth.classic.sim.gap.discovery:
+    timeout: 120
```
This test took 50 seconds on this PR latest CI run, and your own example output was 115s. I think it would be better to have quite a bit more margin here, to account for more loaded runners.
```suggestion
    timeout: 300
```

jhedberg commented:
Heads-up: #116447 proposes a twister sidecar (per the merged #113253 framework) as the host-side orchestration for this simulation framework, and it carries the Python simulation backend from this PR — your commit, trimmed to the Python modules, with authorship and sign-off preserved and a note in the commit message explaining the trim. Please shout if you'd prefer that handled differently.

The intent is that this PR and #111840 remain the home of the actual test content: with the sidecar in place, the shell utilities and per-test scripts here would drop out, and each scenario becomes a `sidecar: bumble` declaration in `tests.yaml` with the ztest harness. See the `acl_conn` reference test in #116447 for the shape.
